cip/CIP2016-06-14-Redefine-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,309 @@
+
+* `MAP` types
+** <<regular-map,Regular maps>>
+** `NODE`
+** `RELATIONSHIP`
+* `LIST OF ANY?`
+* `PATH`
+* `STRING`
+* `BOOLEAN`
+* `NUMBER` (`NaN` is treated as the largest number in orderability only, i.e. it is put after positive infinity)
+* `VOID` (i.e. the type of `null`)
+
+To give a concrete example, under this global sort order all Nodes are less than all Strings.
+
+Between values of the same type in the global sort order, orderability defers to comparability except that equality is overriden by equivalence as described below.
+
+The accompanying descending global sort order is the same order in reverse (i.e. it runs from `VOID` to `NODE`).
+
+[[equivalence-def,equivalence]]
+==== Equivalence ====
+
+Equivalence remains mostly unchanged but now can be defined succinctly as being identical to equality except that:
+
+* Any two `null` values are equivalent (both directly or inside nested structures).
+* Any two `NaN` values are equivalent (both directly or inside nested structures).
+* However, `null` and `NaN` values are not equivalent (both directly or inside nested structures).
+
+<<equivalence-def,Equivalence>> is reflexive for all values.
+
+[[conceptual-model]]
+=== Summary of the conceptual model ===
+
+This proposal aims to simplify the conceptual model around equality, comparison, order, and grouping:
+
+- <<comparability-equality,Comparability and equality>> are aligned with each other
+  * <<comparability-con,Comparability>> ensure that any two values of the same leaf type in the <<global-sort-order>> are comparable. Two values of different types are incomparable and values involving `null` are incomparable, too.
+This ensures that `MATCH (n) WHERE n.prop < 42` will never find nodes where `n.prop` is of type `STRING`.
+  * <<equality-con,Equality>> follows natural, literal equality. However, values involving `null` are never equal to any other value. Nested structures are first tested for equality by shape (keys, size) and then their corresponding elements are tested for equality pairwise. This ensures that equality is compatible with interpreting `null` as "unknown" or "could be any value" or "undefined".
+- <<orderability-equivalence>> are aligned with each other
+  * <<orderability-con,Orderability>> follows comparability but additionally defines a <<global-sort-order>> between values of different types and is aligned with equivalence instead of equality, i.e. treats two `null` (resp. `NaN`) values as equivalent.
+  * <<equivalence-con,Equivalence>> is a form of equality that treats `null` (and `NaN`) values as the same value. Equivalence is used in grouping and `DISTINCT` where `null` commonly is interpreted as a category marker for results with missing values instead of as a wildcard for any possible value.
+
+=== Examples
+
+An integer compared to a float
+[source, cypher]
+----
+RETURN 1 > 0.5 // should be true
+----
+
+A string compared to a boolean
+[source, cypher]
+----
+RETURN 'string' \<= true // should be null
```

Comment:
That's needed to stop the asciidoc renderer from turning this into a unicode symbol (which looks off)